### PR TITLE
Example docs

### PR DESCRIPTION
**Pull request by zhangguanheng66**

### CODE REVIEW

File: examples/text_classification/create_datasets.py [hunk 1]
```diff
@@ -6,6 +6,11 @@
 
 from torchtext.datasets import text_classification
 
+r'''
+Once you have the datasets, you can save the datasets for re-use.
```
Comment (cpuhrsch) [inline examples/text_classification/create_datasets.py]: "save them as a list of Tensors"

File: examples/text_classification/model.py [hunk 2]
```diff
@@ -1,5 +1,15 @@
 import torch.nn as nn
 
+r'''
+The model is composed of the embeddingbag layer and the linear layer.
+nn.EmbeddingBag computes the mean of 'bags' of embeddings. Since it
+doesn't instantiate the intermediate embeddings, nn.EmbeddingBag can
```
Comment (cpuhrsch) [inline examples/text_classification/model.py]: I think you mean "since it accumulate the average across the embeddings on the fly"

File: examples/text_classification/model.py [hunk 3]
```diff
@@ -1,5 +1,15 @@
...
+enhance the performance and memory efficiency to process a sequence
+of tensors. Additionally, the text entries here have different lengths.
+nn.EmbeddingBag requires no padding here so this method is much faster
```
Comment (cpuhrsch) [inline examples/text_classification/model.py]: I'd mention this first.

File: examples/text_classification/model.py [hunk 4]
```diff
@@ -15,4 +25,10 @@ def init_weights(self):
         self.fc.bias.data.zero_()
 
     def forward(self, text, offsets):
+        r'''
+        Arguments:
+            text: a bag of text tensors
```
Comment (cpuhrsch) [inline examples/text_classification/model.py]: The "bag" here means a single 1d-tensor. I'd mention that.

File: examples/text_classification/model.py [hunk 5]
```diff
@@ -15,4 +25,10 @@ def init_weights(self):
...
+            offsets: a list of offsets
```
Comment (cpuhrsch) [inline examples/text_classification/model.py]: I'd mention that these are offsets that delimit the individual sequences within the "bag"

File: examples/text_classification/predict.py [hunk 6]
```diff
@@ -6,6 +6,18 @@
 
 
 def predict(text, model, dictionary):
```
Comment (cpuhrsch) [inline examples/text_classification/predict.py]: dictionary here is a misnomer on my end since we ended up standardizing on the legacy "vocab"

File: examples/text_classification/train.py [hunk 7]
```diff
@@ -10,9 +10,24 @@
...
-def generate_batch(batch):
 
+def generate_batch(batch):
+    r'''
+    Since the text entries have different lengths, a custom function
+    generate_batch() is used to generate data batches and offsets,
+    which are compatible with EmbeddingBag. The function is passed
+    to 'collate_fn' in torch.utils.data.DataLoader. The input to
+    'collate_fn' is a list of tensors with the size of batch_size,
+    and the 'collate_fn' function packs them into a mini-batch.
```
Comment (cpuhrsch) [inline examples/text_classification/train.py]: How does in pack them? In particular, how is "offsets" created.

File: examples/text_classification/train.py [hunk 8]
```diff
@@ -27,7 +42,24 @@ def generate_offsets(data_batch):
     return text, offsets, cls
 
 
+r'''
+torch.utils.data.DataLoader is recommended for PyTorch domain libraries.
```
Comment (cpuhrsch) [inline examples/text_classification/train.py]: I don't think it's specific to domain libraries.

File: examples/text_classification/iterable_train.py [hunk 9]
```diff
@@ -15,8 +15,32 @@
 
 from tqdm import tqdm
 
+r"""
+This example shows how to build an iterable dataset from the iterator. A
+new API get_csv_iterator() is used to read CSV file for the data. An abstract
```
Comment (cpuhrsch) [inline examples/text_classification/iterable_train.py]: I suggest we write "The function get_csv_iterator" instead of "A new API", since API typically refers to sets of functions.

File: examples/text_classification/iterable_train.py [hunk 10]
```diff
@@ -25,7 +49,22 @@ def generate_batch(batch):
...
+r"""
+torch.utils.data.DataLoader is recommended for PyTorch users to load data.
+We use DataLoader here to load datasets and send it to the train()
+and text() functions.
+"""
+
+
 def train(lr_, num_epoch, data_):
+    r"""
+    Here we use SGD optimizer to train the model.
```
Comment (cpuhrsch) [inline examples/text_classification/iterable_train.py]: We still use different optimization approaches across the tutorial, train.py and iterable_train.py. Could you look into unifying them?

File: examples/text_classification/iterable_train.py [hunk 11]
```diff
@@ -72,6 +115,18 @@ def test(data_):
 
 
 def get_csv_iterator(data_path, ngrams, vocab, start=0, num_lines=None):
+    r"""
+    Generate an iterator to read CSV file.
```
Comment (cpuhrsch) [inline examples/text_classification/iterable_train.py]: Add a description of the yielded values.

File: examples/text_classification/iterable_train.py [hunk 12]
```diff
@@ -93,6 +148,15 @@ def iterator(start, num_lines):
 
 
 class Dataset(torch.utils.data.IterableDataset):
```
Comment (cpuhrsch) [inline examples/text_classification/iterable_train.py]: I think "_setup_iterator" will need quite a bit of documentation.

File: examples/text_classification/iterable_train.py [hunk 13]
```diff
@@ -128,6 +192,9 @@ def __iter__(self):
 
 
 def count(data_path):
+    r"""
+    return the total numerbers of text entries and labels.
```
Comment (cpuhrsch) [inline examples/text_classification/iterable_train.py]: nit: "total number of"

File: examples/text_classification/model.py [hunk 14]
```diff
@@ -1,5 +1,19 @@
 import torch.nn as nn
 
+r"""
+The model is composed of the embeddingbag layer and the linear layer.
+
+nn.EmbeddingBag computes the mean of 'bags' of embeddings. The text
+entries here have different lengths. nn.EmbeddingBag requires no
```
Comment (cpuhrsch) [inline examples/text_classification/model.py]: Why does it not require any padding?

File: examples/text_classification/predict.py [hunk 15]
```diff
@@ -5,9 +5,21 @@
 from torchtext.datasets.text_classification import text_normalize
 
 
-def predict(text, model, dictionary):
+def predict(text, model, vocab):
```
Comment (cpuhrsch) [inline examples/text_classification/predict.py]: I think this example might not work anymore, since we moved "text_normalize" into the get_tokenizer abstraction

File: examples/text_classification/train.py [hunk 16]
```diff
@@ -10,9 +10,30 @@
 
 from model import TextSentiment
 
+r"""
+This file shows the training process of the text sentiment model.
```
Comment (cpuhrsch) [inline examples/text_classification/train.py]: This works on all text classification datasets not just those relevant to sentiment.

File: examples/text_classification/create_datasets.py [hunk 17]
```diff
@@ -6,6 +6,11 @@
 
 from torchtext.datasets import text_classification
 
+r"""
+Once you have the datasets, you can save them as a list of tensors
+for re-use. Here is an example to load/save text_classification datasets.
```
Comment (cpuhrsch) [inline examples/text_classification/create_datasets.py]: create and save the datasets. they can be loaded later on in other projects.

File: examples/text_classification/iterable_train.py [hunk 18]
```diff
@@ -15,8 +15,32 @@
 
 from tqdm import tqdm
 
+r"""
+This example shows how to build an iterable dataset from the iterator. The
+get_csv_iterator() function is used to read CSV file for the data. An abstract
```
Comment (cpuhrsch) [inline examples/text_classification/iterable_train.py]: nit: read a CSV file

File: examples/text_classification/iterable_train.py [hunk 19]
```diff
@@ -25,15 +49,30 @@ def generate_batch(batch):
...
+    """
     data = DataLoader(
         data_,
         batch_size=batch_size,
         collate_fn=generate_batch,
         num_workers=args.num_workers,
         pin_memory=True)
     optimizer = torch.optim.SGD(model.parameters(), lr=lr_)
-    scheduler = torch.optim.lr_scheduler.ExponentialLR(optimizer, args.lr_gamma)
+    scheduler = torch.optim.lr_scheduler.StepLR(optimizer, 1, gamma=args.lr_gamma)
```
Comment (cpuhrsch) [inline examples/text_classification/iterable_train.py]: are the default argument names set to sane values for this?

Also, could you add some more docs to the arguments for these scripts?

In particular you use the [help keyword](https://docs.python.org/3/library/argparse.html#adding-arguments) for this. You should then also include the default in that help strings.

File: examples/text_classification/iterable_train.py [hunk 20]
```diff
@@ -145,7 +220,7 @@ def count(data_path):
     parser.add_argument('--num-epochs', type=int, default=3)
     parser.add_argument('--embed-dim', type=int, default=128)
     parser.add_argument('--batch-size', type=int, default=64)
-    parser.add_argument('--lr', type=float, default=64.0)
+    parser.add_argument('--lr', type=float, default=4.0)
```
Comment (cpuhrsch) [inline examples/text_classification/iterable_train.py]: Add docstrings via the 'help' keyword to arguments like this.
